nixos/modules/services/networking/wireguard.nix

```diff
@@ -0,0 +1,165 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.networking.wireguard;
+
+  kernel = config.boot.kernelPackages;
+
+  # interface options
+
+  interfaceOpts = { name, ... }: {
+
+    options = {
+
+      ip = mkOption {
+        example = "192.168.2.1";
+        type = types.str;
+        description = "The ip address added to the interface.";
+      };
+
+      privateKey = mkOption {
+        example = "yAnz5TF+lXXJte14tji3zlMNq+hd2rYUIgJBgB3fBmk=";
+        type = types.str;
+        description = "Private key used by the interface.";
+      };
+
+      presharedKey = mkOption {
+        default = null;
+        example = "rVXs/Ni9tu3oDBLS4hOyAUAa1qTWVA3loR8eL20os3I=";
+        type = with types; nullOr str;
+        description = "Preshared key key used by the interface.";
+      };
+
+      listenPort = mkOption {
+        default = null;
+        type = with types; nullOr int;
+        example = 12345;
+        description = "Port used by the interface.";
+      };
+
+      peers = mkOption {
+        default = [];
+        description = "Peers linked to the interface.";
+        type = with types; listOf (submodule peerOpts);
+      };
+
+    };
+
+  };
+
+  # peer options
+
+  peerOpts = {
+
+    options = {
+
+      publicKey = mkOption {
+        example = "xTIBA5rboUvnH4htodjb6e697QjLERt1NAB4mZqp8Dg=";
+        type = types.str;
+        description = "Public key of the peer.";
+      };
+
+      allowedIPs = mkOption {
+        example = [ "10.192.122.3/32" "10.192.124.1/24" ];
+        type = with types; listOf str;
+        description = "IPs allowed by the peer.";
+      };
+
+      endpoint = mkOption {
+        default = null;
+        example = "test.wireguard.io:18981";
+        type = with types; nullOr str;
+        description = "Endpoint of the peer.";
+      };
+
+      persistentKeepalive = mkOption {
+        default = null;
+        type = with types; nullOr int;
+        example = 12345;
+        description = "Keepalive seconds.";
```

Comment:
This is an appalling description. Did you even read the man page of `wg(8)`? Here's something better:

This is optional and is by default off, because most users will not need it. It represents, in seconds, between 1 and 65535 inclusive, how often  to  send  an  authenticated empty packet to the peer, for the purpose of keeping a stateful firewall or NAT mapping valid persistently. For example, if the interface very rarely sends traffic, but  it  might  at anytime  receive traffic  from  a  peer,  and  it  is behind NAT, the interface might benefit from having a persistent keepalive interval of 25 seconds; however, most users will not need this.